src/mlpack/methods/lsh/lsh_search_impl.hpp

```diff
@@ -331,12 +333,423 @@ void LSHSearch<SortPolicy>::BaseCase(const size_t queryIndex,
+  std::vector<bool> check(numProj);
+
+  for (size_t i = 0; i < A.size(); ++i)
+  {
+    // Check that we only use valid dimensions. If not, vector is not valid.
+    if (A[i] >= 2 * numProj)
+      return false;
+
+    // Check that we only see each dimension once. If not, vector is not valid.
+    if (check[A[i] % numProj ] == 0)
+      check[A[i] % numProj ] = 1;
+    else
+      return false;
+  }
+  return true;
+}
+*/
+
+// Returns the score of a perturbation vector generated by perturbation set A.
+// The score of a pertubation set (vector) is the sum of scores of the
+// participating actions.
+inline double perturbationScore(const std::vector<bool>& A,
+                                const arma::vec& scores)
+{
+  double score = 0.0;
+  for (size_t i = 0; i < A.size(); ++i)
+    if (A[i])
+      score += scores(i); // add scores of non-zero indices
+  return score;
+}
+
+// Inline function used by GetAdditionalProbingBins. The vector shift operation
+// replaces the largest element of a vector A with (largest element) + 1.
+// Returns true if resulting vector is valid, otherwise false.
+inline bool perturbationShift(std::vector<bool>& A)
+{
+  size_t maxPos = 0;
+  for (size_t i = 0; i < A.size(); ++i)
+    if (A[i] == 1) // marked true
+      maxPos=i;
+  
+  if ( maxPos + 1 < A.size()) // otherwise, this is an invalid vector 
+  {
+    A[maxPos] = 0;
+    A[maxPos+1] = 1;
+    return true; // valid
+  }
+  return false; // invalid
+}
+
+// Inline function used by GetAdditionalProbingBins. The vector expansion
+// operation adds the element [1 + (largest_element)] to a vector A, where
+// largest_element is the largest element of A. Returns true if resulting vector
+// is valid, otherwise false.
+inline bool perturbationExpand(std::vector<bool>& A)
+{
+  // Find the last '1' in A
+  size_t maxPos = 0;
+  for (size_t i = 0; i < A.size(); ++i)
+    if (A[i]) // marked true
+      maxPos = i;
+
+  if ( maxPos + 1 < A.size()) // otherwise, this is an invalid vector
+  {
+    A[maxPos + 1] = 1;
+    return true;
+  }
+  return false;
+
+}
+
+// Return true if perturbation set A is valid. A perturbation set is invalid if
+// it contains two (or more) actions for the same dimension or dimensions that
+// are larger than the queryCode's dimensions.
+inline bool perturbationValid(const std::vector<bool>& A,
+                              const size_t numProj)
+{
+  // Use check to mark dimensions we have seen before in A. If a dimension is
+  // seen twice (or more), A is not a valid perturbation.
+  std::vector<bool> check(numProj);
+
+  if (A.size() > 2 * numProj)
+    return false; // This should never happen.
+
+  // Check that we only see each dimension once. If not, vector is not valid.
+  for (size_t i = 0; i < A.size(); ++i)
+  {
+    // Only check dimensions that were included.
+    if (!A[i])
+      continue;
+
+    // If dimesnion is unseen thus far, mark it as seen.
+    if ( check[i % numProj] == false )
```

Comment:
Extra spaces on the inside of the parentheses; we should go with `if (check[i % numProj] == false)`.